include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,408 @@
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+#define _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+
+#if _ONEDPL_BACKEND_SYCL
+
+#include <type_traits>
+#include <cstddef>
+#include <cstdint>
+#include <utility>
+#include <algorithm>
+
+#include <oneapi/dpl/pstl/algorithm_fwd.h>
+#include <oneapi/dpl/pstl/parallel_backend.h>
+#include <oneapi/dpl/pstl/hetero/utils_hetero.h>
+
+#include <oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h>
+#include <oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h>
+#include <oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace internal
+{
+
+// This function is responsible for performing an exclusive segmented scan between work items in shared local memory.
+// Each work item passes __delta_local_id which is the distance to the closest lower indexed work item that
+// processes a segment end. A SLM accessor that is twice the work-group size must be provided.
+// Example across eight work items with addition as the binary operation :
+// -----------------------------
+// Local ID    : 0 1 2 3 4 5 6 7
+// Accumulator : 2 3 1 3 0 5 2 1
+// Delta       : 0 1 2 3 0 1 0 0
+// -----------------------------
+// Return Vals : 0 2 5 6 9 0 5 2
+// -----------------------------
+// __wg_segmented_scan is a derivative work and the reason for the additional copyright notice.
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
```

Comment:
Don't understand why it's inline

Reply:
I agree there is no need for this to be inline. I have removed it.